pcdsdevices/epics/state.py

```diff
@@ -269,6 +292,83 @@ def statesrecord_class(classname, *states, doc="", has_ioc=False):
+    """
+    Status produced by state request
+
+    Parameters
+    ----------
+    device : obj
+        Device with `value` attribute that returns a state string
+
+    desired_state : str
+        Requested state
+
+    timeout : float, optional
+        The default timeout to wait to mark the request as a failure
+
+    settle_time : float, optional
+        Time to wait after completion until running callbacks
+
+    poll_rate : float, optional
+        Rate to keep requesting state from device
+    """
+    def __init__(self, device, desired_state,
+                 timeout=None, settle_time=None, poll_rate=0.1):
+        #Store device attributes
+        self.device = device
+        self.current_state = device.value
+        self.desired_state = desired_state
+        #Immediate success?
+        done = self.current_state == self.desired_state
+        #Start timeout thread
+        super().__init__(timeout=timeout, settle_time=settle_time,
+                         done=done,success=done)
+        #Start state thread
+        thread = threading.Thread(target=self._wait_for_state,
+                                  daemon=True, kwargs={'timeout'  : timeout,
+                                                       'poll_rate': poll_rate})
+        self._state_thread = thread
+        self._state_thread.start()
+
+
+    def _wait_for_state(self,*args,  timeout=None, poll_rate=0.1, **kwargs):
+        """
+        Wait for the state to reach the requested value
+        """
+        try:
+            #Starting time and value
+            current_value   = self.device.value
+            expiration_time = time.time() + timeout if timeout else None
+            #Repeatedly check value
+            while not current_value == self.desired_state:
+                logger.info("Waiting for %s to be at at state %s",
+                            self.device.name, self.desired_state)
+                #Wait for update
+                time.sleep(poll_rate)
+                #Increase time as not to spam EPICS
+                if poll_rate < 0.1:
+                    poll_rate *= 2 #Logarithmic back-off
+                #Check for new value
+                current_value = self.device.value
+                #Timeout exit condition
+                if expiration_time is not None and time.time() > expiration_time:
+                    raise TimeoutError("Device {} failed to reach state {} after {} "
+                                       "seconds. Current value is {}"
+                                       "".format(self.device.name,
+                                                 self.desired_state,
+                                                 current_value))
+            #Report success
+            self._finished(success=True)
+
+        #Catch exceptions and log
+        except Exception as exc:
+            logger.error(exc)
+
+        #Make sure thread gets garbage collected
+        finally:
+            self._state_thread = None
```

Comment:
I don't think looping and polling is a good approach for this. Maybe instead we can subscribe a function to SUB_STATE and mark the status as finished from inside the callback once we reach the correct state.

Reply:
Thought about that. Just have to investigate if there is a good way to clear the callback at the end of thread